Add serialize bilibili json component.

H4M5TER opened the pull request.

H4M5TER committed:
:construction: (worker): json parse progress.

H4M5TER committed:
:construction: (worker): json parse progress.

H4M5TER committed:
:construction: (serialize): DANMU_MSG done.

H4M5TER committed:
:art: (serialize): Use marcos instead of if for avaibility checking.

cqjjjzr commented on CMakeLists.txt:
```diff
@@ -7,6 +7,16 @@ set(CMAKE_CXX_STANDARD 17)
 include(conan.cmake)
 conan_check(REQUIRED)
 
+conan_cmake_run(REQUIRES
```
与当前分支冲突，Merge时注意删掉这一段

H4M5TER replied:
为什么这块你要写下面

cqjjjzr commented on src/worker/borrowed_buffer.h:
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+class borrowed_buffer
```
建议把这一段移到json里面
并且我觉得可以specialize一下，改名为例如`borrowed_message`，因为后面还可能往里面加入routing key的代码

H4M5TER replied:
目前bili_json.cpp里已经有一个borrowed_message继承borrowed_buffer了 这种做法不就是你当初想要的虚函数调用吗
至于borrowed_message需不需要整个对外公开 视后面计划的调用形式而定
我目前感觉依然可以包装起来 即使未来有写入除protobuf外的东西的需求

H4M5TER replied:
目前的需求: 在外界调用serialize获得borrowed_buffer 将负载和routingkey一并写入外界的存储 通过负载计算CRC32
目前的实现: 在外界通过borrowed_buffer的虚函数表将"一块内存"写入外界的存储

需要补充校验码相关的实现

cqjjjzr replied:
CRC32的计算建议使用原JSON数据，在`ParseInsitu`之前计算

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -0,0 +1,263 @@
+    }
+
+// 这个库似乎没有提供指定初始化容量的构造函数
+robin_hood::unordered_map<string, function<void(const Document&, RoomMessage*, Arena*)>> command;
```
map哪来的初始化容量¿

H4M5TER replied:
所以我一开始打算直接把函数写在map的初始化里 希望能确保固定的内存分配 但是后来忘记因为什么转到宏去了 不过这个注释还留着

H4M5TER replied:
经过讨论: 介于command是只对本cpp文件内可访问的 所以放弃const以使用emplace_back是可以接受的

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -0,0 +1,263 @@
+    // 以下变量类型均为 const rapidjson::GenericArray::Iterator&
+    // 如何能够使用++iter并不产生内存开销？
+    ASSERT_TRACE_THROW(document["info"].IsArray())
+    auto const& info_iter = document["info"].Begin();
```
这里为何是iterator而非数组本身

H4M5TER replied:
说实话我并不是很懂rapidjson的最佳实践 譬如document["info"]和document["info].GetArray()有什么区别?
另外iterator我觉得也是可以接受的 毕竟数组里同一个位置的数据需要复用若干次 如果能每次手动++iter就感觉不错 就是不知道存一个iter会不会有内存碎片了

cqjjjzr replied:
你可以存指向Array的引用，另外或许你不应该把iter保存成引用..

H4M5TER replied:
尝试性地重构了一下

H4M5TER replied:
经过讨论 GenericArray::operator[]是和使用iterator开销相近的 可以直接切换使用

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -0,0 +1,263 @@
+    ASSERT_TRACE((user_info_iter + 3)->IsBool() && (user_info_iter + 4)->IsBool())
+    // 这两个字段分别是月费/年费会员
+    // 都为假时无会员 都为真时报错
+    if ((user_info_iter + 3)->GetBool() == (user_info_iter + 4)->GetBool())
```
这里不太清晰 建议写成
```
if (!3 && !4) 无
else if (!3 && 4) 年费
else if (3 && !4) 月费
else error
```

H4M5TER replied:
在每一处添加了备注

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -0,0 +1,263 @@
+    }
+    else if (5000 == (user_info_iter + 5)->GetInt())
+    {
+        user_info->set_regular_user(false);
```
建议单行酌情加花括号

H4M5TER replied:
这里主要是错误处理还没写 前面那些写完的我都统一整改了
顺便一提 不知道为什么 前面写完的用ASSERT_TRACE 最后明明是else却不缩进 非要大括号 但是大括号又太刷kpi了 所以就同一个缩进了 你有什么解决方案吗

cqjjjzr replied:
我觉得你ASSERT_TRACE有点问题
我觉得应该写成
```cpp
#define ASSERT_TRACE(expr) \
if (!(expr)) {
// LOG ERROR
return nullptr; // 或同等类型
}
```

H4M5TER replied:
类型检查已经重构

H4M5TER committed:
:recycle: (serialize): Refactor type check and error handling.

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -31,156 +32,150 @@ namespace vNerve::bilibili
+    ASSERT_TRACE(document["info"].IsArray())
+    // 数组数量不对是结构性错误
+    // 但b站很有可能在不改变先前字段的情况下添加字段
+    // 或许应改成document["info"].MemberCount() >= 15
```
的确如此

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -31,156 +32,150 @@ namespace vNerve::bilibili
     // uid
-    user_info->set_uid(user_info_iter->GetUint64());
-    ASSERT_TRACE((user_info_iter + 1)->IsString())
+    embeded_user_info->set_uid(user_info[0].GetUint64());
```
typo: embedded

H4M5TER committed:
:pencil2: (serialize): Fix embedded typo.

H4M5TER committed:
:ok_hand: (serialize): Change the criteria of broken bilibili json structure.

H4M5TER committed:
:construction: (serialize): DANMU_MSG complete.

Complete missing fields.
Add non-structural error handling.
Clean some comment.

H4M5TER committed:
:pencil2: (serialize): Chore and typos.

H4M5TER committed:
:construction: (serialize): Superchat in progress.

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -0,0 +1,466 @@
+    return _parse_context.get();
+}
+
+const borrowed_buffer* serialize_buffer(char* buf, const unsigned int& room_id)
```
这里为什么是 `int&`

H4M5TER replied:
传入不会更改的参数不就应该是const&吗

cqjjjzr replied:
`int` 等基础类型不需要。直接复制的开销甚至可能小于复制一个64位指针。

H4M5TER replied:
了解

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -150,11 +150,14 @@ CMD(DANMU_MSG)
     // liver user id
-    ASSERT_TRACE(medal_info[3].IsUint())  // IsUint()->GetUint64()可疑
-    embedded_medal_info->set_streamer_uid(medal_info[3].GetUint64());
+    // 弹幕没有牌子属于的用户的uid的字段
```
牌子属于的用户的Uid指的是主播UID？那可能确实无

H4M5TER replied:
SC有 targetid就是

cqjjjzr replied:
那可能需要oneof

H4M5TER replied:
同下下

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -206,9 +209,179 @@ CMD(DANMU_MSG)
-    auto const& _user_info = _data["user_info"];
+    // TODO: 补充SC中的字段
+
+    // 以下变量均为 rapidjson::GenericArray ?
```
难道不是`GenericObject`?

H4M5TER replied:
复制错了

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -206,9 +209,179 @@ CMD(DANMU_MSG)
+
+    // user_info
+    // uid
+    ASSERT_TRACE(data.HasMember("uid"))
```
强烈建议使用 `FindMember` 函数避免多次寻找。应通过判断 `iterator` 与 `MemberEnd` 确定存在性

H4M5TER replied:
请详述“多次寻找”

cqjjjzr replied:
`GenericObject` 内部类似于 `unordered_map<string, GenericValue>`。这段代码会三次利用键寻找值。而改为 `FindObject` 仅会发生一次。

H4M5TER replied:
能不能科普一下为什么会三次查询

H4M5TER replied:
另外MemberEnd是开销极小是吗

cqjjjzr replied:
分别为 `HasMember` 调用和两次 `operator[]` 调用，产生三次查询。  
可以理解为内联后几乎无开销。

H4M5TER replied:
了解

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -206,9 +209,179 @@ CMD(DANMU_MSG)
+    // phone_verified
+    // SC似乎没有这些字段
+    // 但是赠送礼物的理应可以视为正常用户 而不是默认值的非正常用户
+    embedded_user_info->set_regular_user(true);
```
建议未来将 `regular_user` 这几个字段用 `oneof` 保存（oneof 有微小的开销

H4M5TER replied:
可选字段都应改为用oneof保存 transmitter更改还未commit

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -206,9 +209,179 @@ CMD(DANMU_MSG)
+    ASSERT_TRACE(user_info["face"].IsString())
+    embedded_user_info->set_avatar_url(user_info["face"].GetString(), user_info["face"].GetStringLength());
+    // user_info["face_frame"] 舰长框
+    // 没有对应的字段
```
暂时不需要

H4M5TER replied:
需要根据下游需求决定

cqjjjzr commented on src/worker/bili_json.cpp:
```diff
@@ -206,9 +209,179 @@ CMD(DANMU_MSG)
+    auto embedded_user_message = Arena::CreateMessage<live::UserMessage>(arena);
+    auto embedded_user_info = Arena::CreateMessage<live::UserInfo>(arena);
+    auto embedded_gift = Arena::CreateMessage<live::GiftMessage>(arena);
+    // 礼物似乎没有牌子信息
```
默认 `null` 即可

H4M5TER replied:
直接不set_allocated_medal不可以吗

cqjjjzr replied:
可以

H4M5TER committed:
:art: (serialize): Move the serialize detail to the end of file.

H4M5TER committed:
:recycle: (serialize): Refactor borrowed message.

Borrowed message owns crc32 and routing key now.
To modify it in transfer function, refactor borrowed message.

cqjjjzr commented:
参见 vNerve/vNerveTransmitter#2

H4M5TER committed:
:wrench: (cmake): Fix broken CMakeLists.txt.

H4M5TER committed:
:construction: (serialize): Introduce crc32 and routing_key.

H4M5TER committed:
:bug: (serialize): Fix bad codes caused by regex.

H4M5TER committed:
:arrow_up: (global): Update compiled proto files.

H4M5TER committed:
:bug: (serialize): Fix missing declaration of ~parse_context().

H4M5TER committed:
Merge remote-tracking branch 'refs/remotes/vNerve/master'

Conflicts:
	CMakeLists.txt
	proto
	src/worker/bili_packet.cpp
